File: pat1034.py
"""
    思路：采用DFS对连通分量进行遍历，输出结果
    注：为什么采用hnt_info？
        因为python不像c++有引用类型，故采取字典类型使其在递归过程中保存变量值

    3，4，5测试点因为输入数据耗时问题，不能通过

"""


N,K = [int(x) for x in input().split()]
people = []
weight = [0] * N
G = [[0 for i in range(N)] for j in range(N)]   #   初始一个N*N，值为0的邻接矩阵
visited = [False] * N
Gang = {}
hnt_info = {}       #   存储黑帮首领head，成员数量numMember, 总边权totalValue


def DFS(nowVisit, hnt_info):
    
    hnt_info["numMember"] += 1
    visited[nowVisit] = True
    if weight[nowVisit] > weight[hnt_info["head"]]:
        hnt_info["head"] = nowVisit
    for i in range(len(people)):
        if G[nowVisit][i] > 0:
            hnt_info["totalValue"] += G[nowVisit][i]
            G[nowVisit][i] ,G[i][nowVisit] = 0,0    #   删除这条边，防止回头
            if visited[i] == False:
                DFS(i,hnt_info)

def DFSTravel():
    for i in range(len(people)):
        if visited[i] == False:
            hnt_info["head"] = i
            hnt_info["numMember"] = 0
            hnt_info["totalValue"] = 0
            DFS(i, hnt_info)
            if hnt_info["numMember"] > 2 and hnt_info["totalValue"] > K:
                Gang[people[hnt_info["head"]]] = hnt_info["numMember"]



def nameToId(name):
    if name in people:
        return people.index(name)
    else:
        people.append(name)
        return len(people) - 1


for i in range(N):
    n1,n2,t = input().split()
    t = int(t)
    id1 = nameToId(n1)
    id2 = nameToId(n2)
    weight[id1] += t
    weight[id2] += t
    G[id1][id2] += t
    G[id2][id1] += t


DFSTravel()

print(len(Gang))
# 对结果进行字母表排序
for k in sorted(Gang):
    print(k + " " + str(Gang[k]))

File: pat1034.cpp
/*
	思路：1.对name进行转化，如"AAA"对应0，0对应"AAA"
		  2.对各个连通变量进行dfs遍历，对满足要求的结果进行保存

*/


#include<iostream>
#include<string>
#include<map>
using namespace std;

map<string, int> nameToId, Gang;
map<int, string> idToName;
int N, K;
int numPerson = 0;
int weight[2010] = { 0 };
int G[2010][2010] = { 0 };
bool visited[2010] = { false };

void dfs(int nowVisit, int& head, int& numMember, int& totalValue) {
	numMember++;
	visited[nowVisit] = true;
	if (weight[nowVisit] > weight[head]) {
		head = nowVisit;
	}
	for (int i = 0; i < numPerson; i++) {
		if (G[nowVisit][i] > 0) {
			totalValue += G[nowVisit][i];
			G[nowVisit][i] = G[i][nowVisit] = 0;
			if (visited[i] == false) {
				dfs(i, head, numMember, totalValue);
			}
		}

	}

}

void dfsTravel() {
	for (int i = 0; i < numPerson; i++) {
		
		if (visited[i] == false) {
			int head = i, numMember = 0, totalValue = 0;
			dfs(i, head, numMember, totalValue);
			if (numMember > 2 && totalValue > K) {
				Gang[idToName[head]] = numMember;
			}
		}
	}
}

int transfer(string name) {
	if (nameToId.find(name) != nameToId.end()) {
		return nameToId[name];
	}
	else {
		nameToId[name] = numPerson;
		idToName[numPerson] = name;
		return numPerson++;
	}
}
 
int main() {

	cin >> N >> K;
	for (int i = 0; i < N; i++) {
		string n1, n2;
		int t;
		cin >> n1 >> n2 >> t;
		int id1 = transfer(n1);
		int id2 = transfer(n2);
		weight[id1] += t;
		weight[id2] += t;
		G[id1][id2] += t;
		G[id2][id1] += t;

	}
	dfsTravel();
	cout << Gang.size() << endl;
	map<string, int>::iterator it;
	for (it = Gang.begin(); it != Gang.end(); it++) {
		cout << it->first << " " << it->second << endl;
	}



	return 0;
}